File: association_rules/cs/ar.cs
using System;
using System.Collections.Generic;
using System.Linq;

// Structure to represent an itemset
struct Itemset
{
    public List<int> Items;
}

// Structure to represent a frequent itemset
struct FrequentItemset
{
    public Itemset Itemset;
    public int Support;
}

class Program
{
    // Function to check if itemset B is a subset of itemset A
    static bool IsSubset(Itemset A, Itemset B)
    {
        return B.Items.All(itemB => A.Items.Contains(itemB));
    }

    // Function to check if an itemset is frequent
    static bool IsFrequent(Itemset itemset, List<Itemset> transactions, int minSupport)
    {
        int count = 0;
        foreach (var transaction in transactions)
        {
            bool found = true;
            foreach (var item in itemset.Items)
            {
                if (!transaction.Items.Contains(item))
                {
                    found = false;
                    break;
                }
            }
            if (found)
            {
                count++;
            }
        }
        return count >= minSupport;
    }

    // Function to generate frequent itemsets
    static void GenerateFrequentItemsets(List<Itemset> transactions, int numItems, int minSupport)
    {
        List<Itemset> itemsets = new List<Itemset>();

        // Initialize with single items
        for (int i = 1; i <= numItems; i++)
        {
            Itemset itemset = new Itemset { Items = new List<int> { i } };
            itemsets.Add(itemset);
        }

        // Generate frequent itemsets
        for (int k = 2; k <= numItems; k++)
        {
            List<Itemset> newItems = new List<Itemset>();

            for (int i = 0; i < itemsets.Count; i++)
            {
                for (int j = i + 1; j < itemsets.Count; j++)
                {
                    // Create a candidate itemset
                    Itemset candidate = new Itemset();
                    candidate.Items = new List<int>(itemsets[i].Items);
                    candidate.Items.Add(itemsets[j].Items.Last());

                    // Check if the candidate itemset is frequent
                    if (IsFrequent(candidate, transactions, minSupport))
                    {
                        newItems.Add(candidate);
                    }
                }
            }

            // Copy the new itemsets
            foreach (var newItem in newItems)
            {
                itemsets.Add(newItem);
            }
        }

        // Print frequent itemsets
        foreach (var itemset in itemsets)
        {
            Console.Write("Itemset: ");
            foreach (var item in itemset.Items)
            {
                Console.Write(item + " ");
            }
            Console.WriteLine();
        }
    }

    static void Main()
    {
        int numTransactions = 5;
        int numItems = 4;
        int minSupport = 2;

        // Create a sample dataset
        List<Itemset> transactions = new List<Itemset>
        {
            new Itemset { Items = new List<int> { 1, 2, 3 } },
            new Itemset { Items = new List<int> { 2, 3 } },
            new Itemset { Items = new List<int> { 1, 3, 4 } },
            new Itemset { Items = new List<int> { 1, 3 } },
            new Itemset { Items = new List<int> { 1, 2, 3 } }
        };

        GenerateFrequentItemsets(transactions, numItems, minSupport);
    }
}

File: association_rules/scala/ar.scala
case class Itemset(items: List[Int])

case class FrequentItemset(itemset: Itemset, support: Int)

def isSubset(a: Itemset, b: Itemset): Boolean =
  b.items.forall(a.items.contains)

def isFrequent(itemset: Itemset, transactions: List[Itemset], minSupport: Int): Boolean = {
  val count = transactions.count(t => isSubset(t, itemset))
  count >= minSupport
}

def generateFrequentItemsets(transactions: List[Itemset], numItems: Int, minSupport: Int): Unit = {
  var itemsets = List.empty[Itemset]

  // Initialize with single items
  for (i <- 1 to numItems) {
    val itemset = Itemset(List(i))
    itemsets ::= itemset
  }

  // Generate frequent itemsets
  for (k <- 2 to numItems) {
    var newItems = List.empty[Itemset]

    for (i <- 0 until itemsets.length) {
      for (j <- (i + 1) until itemsets.length) {
        // Create a candidate itemset
        val candidateItems = itemsets(i).items ++ List(itemsets(j).items.last)
        val candidate = Itemset(candidateItems)

        // Check if the candidate itemset is frequent
        if (isFrequent(candidate, transactions, minSupport)) {
          newItems ::= candidate
        }
      }
    }

    // Copy the new itemsets
    itemsets :::= newItems
  }

  // Print frequent itemsets
  itemsets.foreach(itemset => println("Itemset: " + itemset.items.mkString(", ")))
}

val numTransactions = 5
val numItems = 4
val minSupport = 2

// Create a sample dataset
val transactions = List(
  Itemset(List(1, 2, 3)),
  Itemset(List(2, 3)),
  Itemset(List(1, 3, 4)),
  Itemset(List(1, 3)),
  Itemset(List(1, 2, 3))
)

generateFrequentItemsets(transactions, numItems, minSupport)
